File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/reflections/LoggingReflector.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports System
Imports com.db4o.reflect
Imports com.db4o.reflect.net

Namespace com.db4odoc.f1.reflections
    Public Class LoggingReflector
        Implements Reflector
        Private _arrayHandler As LoggingArray
        Private _parent As Reflector

        Public Sub New()
        End Sub

        Public Overridable Function Array() As ReflectArray Implements Reflector.Array
            If _arrayHandler Is Nothing Then
                _arrayHandler = New LoggingArray(_parent)
            End If
            Return _arrayHandler
        End Function

        Public Overridable Function ConstructorCallsSupported() As Boolean Implements Reflector.ConstructorCallsSupported
            Return True
        End Function

        Public Overridable Function ForClass(ByVal clazz As j4o.lang.Class) As ReflectClass Implements Reflector.ForClass
            Dim rc As ReflectClass = New NetClass(_parent, clazz)
            If rc Is Nothing Then
                Console.WriteLine("ForClass: " + clazz.GetName + " -> ... ")
            Else
                Console.WriteLine("ForClass: " + clazz.GetName + " -> " + (rc.GetName()))
            End If
            Return rc
        End Function

        Public Overridable Function ForName(ByVal className As String) As ReflectClass Implements Reflector.ForName
            Try
                Dim clazz As j4o.lang.Class = j4o.lang.Class.ForName(className)
                Dim rc As ReflectClass = ForClass(clazz)
                If rc Is Nothing Then
                    Console.WriteLine("ForName: " + clazz.GetName + " -> ... ")
                Else
                    Console.WriteLine("ForName: " + clazz.GetName + " -> " + (rc.GetName()))
                End If
                Return rc
            Catch e As j4o.lang.ClassNotFoundException
                Return Nothing
            End Try
        End Function

        Public Overridable Function ForObject(ByVal a_object As Object) As ReflectClass Implements Reflector.ForObject
            If a_object Is Nothing Then
                Return Nothing
            End If
            Dim rc As ReflectClass = _parent.ForClass(j4o.lang.Class.GetClassForObject(a_object))
            If rc Is Nothing Then
                Console.WriteLine("ForObject: " + a_object.ToString + " -> ... ")
            Else
                Console.WriteLine("ForObject: " + a_object.ToString + " -> " + (rc.GetName()))
            End If
            Return rc
        End Function

        Public Overridable Function IsCollection(ByVal claxx As ReflectClass) As Boolean Implements Reflector.IsCollection
            Return False
        End Function

        Public Overridable Sub SetParent(ByVal reflector As Reflector) Implements Reflector.SetParent
            _parent = reflector
        End Sub

        Public Overridable Function DeepClone(ByVal context As Object) As Object Implements Reflector.DeepClone
            Return New LoggingReflector()
        End Function
    End Class
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/reflections/LoggingReflector.cs
/* Copyright (C) 2005   db4objects Inc.   http://www.db4o.com */

using System;

namespace com.db4odoc.f1.reflections
{
	public class LoggingReflector : com.db4o.reflect.Reflector
	{
		private LoggingArray _arrayHandler;
		private com.db4o.reflect.Reflector _parent;

		public LoggingReflector()
		{
		}

		public virtual com.db4o.reflect.ReflectArray Array()
		{
			if (_arrayHandler == null)
			{
				_arrayHandler = new LoggingArray(_parent);
			}
			return _arrayHandler;
		}

		public virtual bool ConstructorCallsSupported()
		{
			return true;
		}

		public virtual com.db4o.reflect.ReflectClass ForClass(j4o.lang.Class clazz)
		{
			com.db4o.reflect.ReflectClass rc = new com.db4o.reflect.net.NetClass(_parent, clazz);
			Console.WriteLine("ForClass: " + clazz+" -> "+(rc== null ? "" : rc.GetName()));    
			return rc;
		}

		public virtual com.db4o.reflect.ReflectClass ForName(string className)
		{
			try
			{
				j4o.lang.Class clazz = j4o.lang.Class.ForName(className);
				com.db4o.reflect.ReflectClass rc = ForClass(clazz);
				Console.WriteLine("ForName: " + clazz+" -> "+(rc== null ? "" : rc.GetName()));    
				return rc;
			}
			catch (j4o.lang.ClassNotFoundException e)
			{
				return null;
			}
		}

		public virtual com.db4o.reflect.ReflectClass ForObject(object a_object)
		{
			if (a_object == null)
			{
				return null;
			}
			com.db4o.reflect.ReflectClass rc = _parent.ForClass(j4o.lang.Class.GetClassForObject(a_object));
			Console.WriteLine("ForObject:" + a_object+" -> "+(rc== null ? "" : rc.GetName()));
			return rc;
		}

		public virtual bool IsCollection(com.db4o.reflect.ReflectClass claxx)
		{
			return false;
		}

		public virtual void SetParent(com.db4o.reflect.Reflector reflector)
		{
			_parent = reflector;
		}

		public virtual object DeepClone(object context)
		{
			return new LoggingReflector();
		}
	}
}

